source-academy/cadet-frontend
# Fix REPL not clearing input on eval

## [Bugfix] Fix REPL not clearing input on eval
Summary: Fixes issue #1349.

### Description
Starting from commit 703510654c6594339b136091d2dddecce2f35638, the REPL (a React Ace component itself) no longer correctly clears the REPL input when it is evaluated - this suggests a bug with one of the dependencies that was upgraded.

v9.1.0 of the `react-ace` package includes commit https://github.com/securingsincity/react-ace/commit/a04be50802a7f3fdc8c3e47c0fe923883c85a368, of which the key change is shown below.

![image](https://user-images.githubusercontent.com/44989315/87249379-5c88e980-c491-11ea-8994-bd1200b70151.png)

This commit introduces the additional condition (L333) `nextProps.value` required for the React Ace component (of which the REPL is one) to set the value of its underlying editor (L339) to the new value `nextProps.value`. L308 ([here](https://github.com/securingsincity/react-ace/blob/a04be50802a7f3fdc8c3e47c0fe923883c85a368/src/ace.tsx#L308)) shows that `nextProps` is in fact just `this.props` and that `nextProps.value` is an optional prop.

However, this change unintentionally breaks the behaviour of the Academy's REPL since we set the REPL's value to `''` (the empty string) to clear the REPL input each time the REPL's inputs are evaluated. Since the empty string in Javascript is falsey, updating the REPL's value to `''` causes `nextProps.value` to evaluate to `false`. This causes the React Ace component to skip updating the value of the underlying Ace editor in `componentDidUpdate`, and as a result the initial input persists in the REPL, as reported in issue #1349.

The REPL behaves correctly if its value is set to some non-falsey value when its input is cleared in the workspace reducer - you can test this by setting `replValue` in L234 below to a non-empty string.

https://github.com/source-academy/cadet-frontend/blob/7563672c0bcb90159bd525268949e0f3ed8b8a8a/src/commons/workspace/WorkspaceReducer.ts#L229-L236

I will (probably) make a PR upstream to change the condition to `nextProps.value === undefined` instead to allow `value` to be an empty string.

### Changelog
- Rollback `react-ace` package from v9.1.1 to v9.0.0 due to commit https://github.com/securingsincity/react-ace/commit/a04be50802a7f3fdc8c3e47c0fe923883c85a368

### Verification instructions
1. Re-install all dependencies with `yarn install` - make sure v9.0.0 of the package `react-ace` is installed (you can check this with `yarn list --pattern react-ace`)
2. Restart the development server, navigate to the Playground, enter some input into the REPL and then eval (Shift-Enter) the REPL input; this should clear the REPL input

Last updated 12 July 2020, 11:45PM
